smarthome.cpp

```diff
@@ -8,82 +8,150 @@
 #include <iostream>
 
 namespace smartHome{
-SmartHome::SmartHome(const std::vector<Room>& rooms)
-    : rooms(rooms) {}
+    SmartHome::SmartHome(const std::vector<Room>& rooms)
+        : rooms(rooms) {}
 
-const std::vector<Room>& SmartHome::getRooms() const {
-    return rooms;
-}
+    const std::vector<Room>& SmartHome::getRooms() const {
+        return rooms;
+    }
 
-void SmartHome::readDataFromFile(const std::string& filename) {
-    std::ifstream file(filename);
+    void SmartHome::readDataFromFile(const std::string& filename) {
```

Comment:
if you have the new derived classes, why not use them here to assign proper class types for your sensors and devices?